src/lib.rs

```diff
@@ -187,9 +187,103 @@ macro_rules! make_guard {
         #[allow(unused)]
         let lifetime_brand = unsafe { $crate::LifetimeBrand::new(&branded_place) };
         let $name = unsafe { $crate::Guard::new(branded_place) };
+        if false {
+            // Prelude for poorman's specialization.
+            #[allow(unused)]
+            use $crate::__private::DefaultGet as _;
+            #[allow(unreachable_code)] {
+                let phony = $crate::__private::Phony;
+                if false {
+                    // Help inference make out that the type param here is that
+                    // of the return type.
+                    return $crate::__private::DefaultGet::get(&phony);
+                }
+                // Guarding against `Phony<!>` itself does not suffice, we may
+                // be dealing with `Phony<(!, !)>`, for instance.
+                //
+                // Observation: the very same mechanism which causes us trouble
+                // yields an `unreachable_code` warning in the following situation:
+                if false {
+                    let _reified_ret = $crate::__private::DefaultGet::get(&phony);
+                    #[forbid(unreachable_code)] {
```

Comment:
I'm aware we end up in a rather silly `#[allow(X)] … #[forbid(X)]` situation, here. I've considered removing the `allow`, but then the lint fires a bunch of times. The way it is done currently, it fires exactly once, so it does yield slightly better DX.